haskell.packages.ghcjs: fix unevaluatable packages

trofi opened the pull request:
`haskell.packages.ghcjs` has a few packages that fails to evaluate by referring non-existent attributes. This turns evaluation attempts into unrecoverable errors.

Before the change an attempt to instantiate all `ghcjs` packages failed as:

    $ nix-instantiate --strict --eval --expr 'with import ./. {}; lib.mapAttrs (n: v: builtins.tryEval (lib.isDerivation v)) haskell.packages.ghcjs'
    error:
       … while evaluating the attribute 'llvmPackages'
       error: attribute 'llvmPackages' missing
       at pkgs/development/haskell-modules/configuration-ghc-8.10.x.nix:11:48:

           10|
           11|   llvmPackages = pkgs.lib.dontRecurseIntoAttrs self.ghc.llvmPackages;
             |                                                ^

The change follows definition of `pkgs/top-level/haskell-packages.nix` and also sets `llvmPackages` to `null`. While at it dropped other references to deleted packages like `stm` pin.

## Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

trofi committed:
haskell.packages.ghcjs: fix `stm` eval

Before the change an attempt to instantiate `stm` package failed as:

    $ nix build --no-link -f. haskell.packages.ghcjs.stm
    error:
       … in the left operand of the update (//) operator

         at /home/slyfox/dev/git/nixpkgs-haskell-updates/pkgs/development/haskell-modules/lib/compose.nix:40:7:

           39|     mkDerivation = drv: (args.mkDerivation drv).override f;
           40|   })) // {
             |       ^
           41|     overrideScope = scope: overrideCabal f (drv.overrideScope scope);

       error: attribute 'stm_2_5_1_0' missing

       at /home/slyfox/dev/git/nixpkgs-haskell-updates/pkgs/development/haskell-modules/configuration-ghcjs.nix:29:21:

           28|   # https://github.com/ghcjs/ghcjs/issues/676
           29|   stm = doJailbreak self.stm_2_5_1_0;
             |                     ^
           30|   exceptions = dontCheck self.exceptions_0_10_7;
       Did you mean stm_2_5_3_0?

cdepillabout commented on pkgs/development/haskell-modules/configuration-ghcjs.nix:
```diff
@@ -18,6 +18,8 @@ self: super:
   inherit (self.ghc.bootPkgs)
     jailbreak-cabal alex happy gtk2hs-buildtools rehoo hoogle;
 
+  llvmPackages = null;
```
Could you add a comment here for why this is necessary?  It is not immediately obvious to me why this would be necessary, or semantically why it would be needed.

trofi replied:
Added a comment around it. AFAIU `llvmPackages` is inherited from `configuration-ghc-8.10.x.nix` (shared across `ghcjs` and upstream `8.10`) which defined it as `llvmPackages = pkgs.lib.dontRecurseIntoAttrs self.ghc.llvmPackages;`. As `ghcjs` does not define any `llvmPackages` that makes `llvmPackages` (and packages that depend on it) attribute in this file unevaluatable without the change to recover from an error.

The change fixed `llvmPackages` evaluation into a `null` and unbreaks evaluation of haskell dependent packages. They will likely still fail to build.

sternenseemann replied:
It may be nicer to `passthru.llvmPackages = null` in ghcjs?

trofi replied:
AFAIU that would still fail the evaluation:

```
nix-repl> pkgs.lib.dontRecurseIntoAttrs null
error:
       … in the left operand of the update (//) operator

         at /home/slyfox/dev/git/nixpkgs-haskell-updates/lib/attrsets.nix:1180:11:

         1179|     attrs:
         1180|     attrs // { recurseForDerivations = false; };
             |           ^
         1181|

       error: value is null while a set was expected
       ```
       
       Would it also need a tweak of `configuration-ghc-8.10.x.nix`?

sternenseemann replied:
I guess… or tweak `lib.dontRecurseIntoAttrs`…

trofi replied:
Added `llvmPackages = null` to `ghcjs` and allowed `recurseIntoAttrs` / `dontRecurseIntoAttrs` to accept `null`.

infinisil replied:
How about `llvmPackages = {}`?

sternenseemann replied:
There is code that explicitly checks if `llvmPackages` is null, so that won't work. I don't care how this is solved, but `llvmPackages = {}` won't work.

trofi replied:
Switched `ghcjs` to `llvmPackages = null;` and added a check to `configuration-ghc-8.10.x.nix` as:

```nix
  # ghcjs does not use `llvmPackages` and exposes `null` attribute.
  llvmPackages =
    if self.ghc.llvmPackages != null
    then pkgs.lib.dontRecurseIntoAttrs self.ghc.llvmPackages
    else null;
```

sternenseemann commented on pkgs/development/haskell-modules/configuration-ghcjs.nix:
```diff
@@ -26,7 +28,6 @@ self: super:
 
   # GHCJS does not ship with the same core packages as GHC.
   # https://github.com/ghcjs/ghcjs/issues/676
-  stm = doJailbreak self.stm_2_5_1_0;
```
This instead needs to be changed to `stm_2_5_3_0`. ghcjs doesn't ship `stm` as a core package, so this would break other packages that depend on it.

trofi replied:
Changed to `stm_2_5_3_0`. Thank you!

trofi committed:
haskell.compiler.ghcjs: expose `llvmPackages` attribute via passthru

Files like `pkgs/development/haskell-modules/configuration-ghc-8.10.x.nix`
assume `ghc` always has an `llvmPackages` attribue. Let's expose `null`
value from `ghcjs` to allow it's propagation.

This fixes package evaluation for `ghcjs` packages.

infinisil commented on lib/attrsets.nix:
```diff
@@ -1167,7 +1167,9 @@ rec {
     # An attribute set to scan for derivations.
     attrs:
-    attrs // { recurseForDerivations = true; };
+    if (attrs != null)
+    then attrs // { recurseForDerivations = true; }
+    else null;
```
I don't think changing `[dont]recurseIntoAttrs` for this is justifiable..

trofi replied:
Dropped `lib/` changes and switched to `llvmPackages = {};`. That fixes eval for me as well.
